### PR TITLE
[WiP] Adds JUnit5 extension for Hoverfly Core integration

### PR DESCRIPTION
**Pull request by lordofthejars**
This is only first implementation to start discussing. When we agree with this basis then I implement the extension for automatic Simulate and automatic Capture.

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,7 +1,6 @@
-#Mon Dec 05 11:54:51 GMT 2016
+#Fri Jun 23 08:01:21 CEST 2017
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-3.1-bin.zip
-org.gradle.daemon=false
\ No newline at end of file
+distributionUrl=https\://services.gradle.org/distributions/gradle-3.1-all.zip
```
Comment (lordofthejars) [inline gradle/wrapper/gradle-wrapper.properties]: This change has been done automatically, I suspect that Gradle itself is fixing some ¿deprecated? paths.

File: junit5/src/main/java/io/specto/hoverfly/junit5/spi/HoverflySimulation.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package io.specto.hoverfly.junit5.spi;
+
+import io.specto.hoverfly.junit.core.Hoverfly;
+import io.specto.hoverfly.junit.core.SimulationSource;
+
+/**
+ * Interface to be implemented for specifying Simulation in Hoverfly instance
+ * For example DSL, url, file, ...
+ */
+public interface HoverflySimulation {
```
Comment (lordofthejars) [inline junit5/src/main/java/io/specto/hoverfly/junit5/spi/HoverflySimulation.java]: Still not used in code.

**Comment by mogronalol** [comment]
Thanks, we will take a look tomorrow.

**Comment by lordofthejars** [comment]
Sure, this is only a first approach to start a discussion, but not a final solution.

**Comment by tommysitu** [comment]
@lordofthejars This is exciting. I have been looking into some JUnit 5 examples, and I got a few questions: 

1. Shall we implement a `HoverflyExtension` which basically mirror what `HoverflyRule` does? So that user don't need to call `importSimulation` and `exportSimulation` directly? 
2. Maybe it should always start Hoverfly once before all tests? So `HoverflyExtention` implements `BeforeAllCallBack` and `AfterAllCallBack` to start/stop Hoverfly and import/export simulations. Then it also implments `BeforeEachCallBack` to reset journal for verification. 
3. I assume that the JUnit 5 extension will be packaged as a separate library, for example, `hoverfly-java-junit5`. It can be imported separately when it is needed.

**Comment by lordofthejars** [comment]
1. I was thinking on creating an extension called `HoverflySimulate`, `HoverflyCapture` and `HoverflyCaptureOrSimulate` so user can choose between them or `HoverflyCore`if they want to have full control.

2. Yes in simulate, capture and simulate and capture will work as you mention. In `HoverflyCore` works as you mention if and only if the field is `static`, if not follows the JUnit5 lifecycle.

3. Correct.

**Comment by mogronalol** [comment]
I think this looks good. My thoughts are:

1. Could we try reproduce all the JUnit rule tests with the JUnit5 extensions, then we can compare the usability and see if anything needs to be refined? 
2. I think we could think about convention over configuration for imports / exports. So for example, it could default to "src/test/resources/simulations/<class_name>/<test-name>" if you don't specify a path. 
3. I agree with Tommy that automatically importing and exporting would be useful, but having one annotation per mode could do this as you suggested. So a `@HoverflyCapture,@HoverflySimulateOrCapture,@HoverflySimulate`
3. If you have a single instance, you should be able to change the mode between tests. The use case of this is wanting to update the data for a single test by changing it to capture mode.

**Comment by lordofthejars** [comment]
Well point 3 are not annotations but extensions. I think that for JUnit5 each extension is a different lifecycle. Then let me work a bit more now that we have the basis.

Point 2, yes I agree I was thinking exactly the same.

Point 4, yes with `HoverflyCoreExtension` this is possible.

**Comment by tommysitu** [comment]
@lordofthejars what you suggested sounds good. Because JUnit 5 supports multiple extensions, then maybe we don't need a separate `HoverflySimulateOrCapture` extension, it will be just: 

```
@ExtendWith({HoverflySimulate.class, HoverflyCapture.class})
```
If the simulation files don't exist, only HoverflyCapture extension will be enabled?

@mogronalol, `@ExtendWith` can be applied to method leve, so it is not a problem for 4.

**Comment by lordofthejars** [comment]
@tommysitu true :D learning new stuff together.

File: junit5/build.gradle [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+archivesBaseName = 'hoverfly-java-junit5'
+
+repositories {
+    mavenCentral()
+}
+
+dependencies {
+
+    compileOnly 'org.junit.jupiter:junit-jupiter-api:5.0.0-M4'
```
Comment (eddumelendez) [inline junit5/build.gradle]: update to M5

File: settings.gradle [hunk 4]
```diff
@@ -0,0 +1 @@
+include 'junit5'
```
Comment (eddumelendez) [inline settings.gradle]: I think `junit-jupiter` can be better

**Comment by lordofthejars** [comment]
@eddumelendez Thanks for the feedback. Yes the version is already updated, but PR was sent before M5 was released. Working some minutes everyday to have something fully workable.

**Comment by lordofthejars** [comment]
I have updated junit version and implemented `capture` and `simulate` mode. Now I would like to discuss which could be the best strategy for `captureOrSimulate`.

**Comment by tommysitu** [comment]
@lordofthejars we will look into this PR in the next couple of days.

File: junit5/build.gradle [hunk 5]
```diff
@@ -0,0 +1,40 @@
+buildscript {
+    repositories {
+        mavenCentral()
+    }
+    dependencies {
+        classpath 'org.junit.platform:junit-platform-gradle-plugin:1.0.0-M4'
```
Comment (eddumelendez) [inline junit5/build.gradle]: change to version 1.0.0-M5

Comment (lordofthejars) [inline junit5/build.gradle]: Thanks mate, I changed everywhere except in the plugin :)

**Comment by tommysitu** [comment]
@lordofthejars I have tried update your PR to use the release version of JUnit 5, and it is working fine. 

I got a question though: 
Do you always need to create a wrapper class for `SimulationSource` in order to pass it to `@HoverflySimulate` annotation?

File: junit5/src/test/java/io/specto/hoverfly/junit5/HoverflyCoreStaticTestInstanceTest.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    @HoverflyCore
+    public static Hoverfly hoverfly;
+
+    @Test
+    public void shouldInjectHoverflyAndStartItAutomatically() {
+        assertThat(hoverfly.getHoverflyInfo()).isNotNull();
+    }
+
+    @Test
+    public void shouldInjectHoverflyParameterBeingTheSameAsStatic(@HoverflyCore(mode = HoverflyMode.CAPTURE) Hoverfly hoverfly) {
```
Comment (tommysitu) [inline junit5/src/test/java/io/specto/hoverfly/junit5/HoverflyCoreStaticTestInstanceTest.java]: So `@HoverflyCore(mode = HoverflyMode.CAPTURE)` is ignored here? If you need to change Hoverfly mode for this method, you would need to call this: `hoverfly.setMode(HoverflyMode.CAPTURE);`

File: junit5/src/test/java/io/specto/hoverfly/junit5/HoverflyCoreTestInstanceTest.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.catchThrowable;
+
+@ExtendWith(HoverflyCoreResolver.class)
+public class HoverflyCoreTestInstanceTest {
+
+    @HoverflyCore
+    public Hoverfly hoverfly;
```
Comment (tommysitu) [inline junit5/src/test/java/io/specto/hoverfly/junit5/HoverflyCoreTestInstanceTest.java]: This is never used.

File: junit5/src/test/java/io/specto/hoverfly/junit5/HoverflyCoreStaticTestInstanceTest.java (hunk 6 above)
Comment (lordofthejars) [reply junit5/src/test/java/io/specto/hoverfly/junit5/HoverflyCoreStaticTestInstanceTest.java]: No, in this case you set hoverfly in `CAPTURE` mode.

File: junit5/src/test/java/io/specto/hoverfly/junit5/HoverflyCoreTestInstanceTest.java (hunk 7 above)
Comment (lordofthejars) [reply junit5/src/test/java/io/specto/hoverfly/junit5/HoverflyCoreTestInstanceTest.java]: Yes in line 21 for example

**Comment by lordofthejars** [comment]
> Do you always need to create a wrapper class for SimulationSource in order to pass it to @HoverflySimulate annotation? Is there a different way to do it?

If you don't relay on default value then yes. Or you can inject `HoverflyCore` and manage yourself in test instead by JUnit 5 extension.

**Comment by tommysitu** [comment]
@lordofthejars I've finished reviewing this PR now. Having tried to migrate some existing tests to use this new feature, I come across a few things that can be improved: 
1. Creating a separate class to wrap simulation source or the Hoverfly config is a bit tedious, and not very intuitive. I would propose this: 
```
@HoverflySimulate(
        configs = @HoverflyConfig(adminPort = 8889, proxyPort = 8890),
        source = @HoverflySimulate.Source(classpath = "test-service.json"))
@ExtendWith(HoverflyExtension.class)
```
which is equivalent to this with HoverflyRule:
```
@ClassRule
public static HoverflyRule hoverflyRule = HoverflyRule.inSimulationMode(
       classpath("test-service.json"),
       configs().proxyPort(8890).adminPort(8889));
```
and for Capture mode:
```
@HoverflyCapture(
        recordFile = "recorded-simulation.json",
        configs = @HoverflyConfig(proxyLocalhost = true, captureAllHeaders = true))
@ExtendWith(HoverflyExtension.class)
```
equivalent to: 
```
 @ClassRule
 public HoverflyRule hoverflyRule = HoverflyRule.inCaptureMode(
        "recorded-simulation.json",
        configs().captureAllHeaders().proxyLocalHost());
```

2. I don't see any use case to start / stop Hoverfly per test method. So I would propose to just implement `BeforeAll` for `HoverflyCoreResolver` and start Hoverfly once for a test class, then implement `BeforeEach` to reset the state of the Hoverfly, such as the simulation, journal log and state store.

3. I don't see any point to allow field inject for Hoverfly using `@HoverflyCore` annotation, if we want to have a fine grain control of Hoverfly, we can do this: 
```
@HoverflyCore(
        configs = @HoverflyConfig(adminPort = 8889, proxyPort = 8890)   // provide start-up time customization
@ExtendWith(HoverflyExtension.class)

@BeforeAll
public static void beforeAll(Hoverfly hoverfly) {
   // configure hoverfly before all tests
}

@BeforeEach
public void beforeEach(Hoverfly hoverfly) {
   // configure hoverfly before each test
}

@Test
public void someTest(Hoverfly hoverfly) {
   // configure hoverfly for this test
}
```
Actually the same applied for  `@HoverflySimulate` and `@HoverflyCapture` as you have implemented `ParameterResolver`, the hoverfly instance can be injected into each test, for instance, to simulate using DSL.

4. Optional: It would be nice to have only one Extension `HoverflyExtention` which encapsulate all the logics in `HoverflyCoreResolver`, `HoverflySimulateResolver` and `HoverflyCaptureResolver`, then switch between different modes by simply using the `@HoverflySimlate` or `@HoverflyCapture` annotations. 

5. I would argue that we need to have another `HoverflySimulateOrCaptureResolver`, what we can do instead is add an extra flag in `@HoverflySimulate` like this:
```
@HoverflySimulate(captureIfNotPresent = true)
```
If no source is provided, it will export the simulation using class name as filename and if a source is provided, use the source name.

**Comment by lordofthejars** [comment]
@tommysitu Great. I won't be able to work until next week. I just say because if you are inhurry, I can give you permissions to my PR so you can write as well.

**Comment by tommysitu** [comment]
That's OK, if you can give me the permission, I will go ahead and chuck in some code as I have some free time recently. Sorry for taking a while to review this as well.

**Comment by lordofthejars** [comment]
Ok, I have sent you an invitation, now you'll be able to push changes to my repo which will be reflected at PR

**Comment by mogronalol** [comment]
Hey, I've checked over this again, and everything looks good to me. JUnit5 is a lot nicer.

**Comment by lordofthejars** [comment]
Looks great. @tommysitu what it is the `autoCapture` flag?

Also hope we can have this released before Devoxx so I can show how to use Hoverfly + JUnit 5 in my presentation.

**Comment by tommysitu** [comment]
@lordofthejars the `autoCapture` flag is used for enabling the `inSimulateOrCaptureMode`. 
I should be able to write up the documentation and release it this week.

**Comment by lordofthejars** [comment]
Great, perfect timing for Devoxx

**Comment by tommysitu** [comment]
Merging this in now. I will do the docs in a separate branch.

**Pull request merged**